### ISSUE TITLE
Multiline column

### ISSUE THREAD

**Issue by bastien-besson** [bytefish/TinyCsvParser#26]
Hi, I wanted to test the different csv parser performance and abilities, and found that there is no dealing with multilines in TinyCsvParser.
I guess it would be possible to change that with an optional relining method before the File.ReadLines method used in ReadFromFile, even if if will of course affect performance. 
Just wanted to share the thoughts.
Cheers and awesome job !

**Comment by bytefish** [comment]
Hi Bastien, thanks for your feedback. Yes I know this is a limitation, which is a limitation. I think I can modify the pipeline a little to use the RFC4180 Tokenizer for tokenizing the data into lines. I might extend it next week, if I find the time. ☺️

**Comment by bytefish** [comment]
I thought about it this morning, and maybe you have something to add. If I would add multiline columns, then I cannot parallelize the Tokenization as it is. Supporting multiline columns would make it necessary to parse the file sequentially and have all cores waiting for new Tokenized data in between.

Of course one could start also start multiple Parsers and parallelize on a per-file basis. What do you think? Maybe I could add an option for multiline support, so the User can decide by himself if to take a performance hit.

**Comment by bastien-besson** [comment]
Hello Philipp,

I agree with you on the necessity of making the multiline support optional,
as the performance hit will clearly be important.

For the process, I saw that the csvHelper project pulls in a buffer of
data,read every character one after another and check for the delimiters ,
quotes and end of lines to "reline" the multilines columns.
For performance optimisation, I thought about maybe chuncking the csv file
and doing kind of the same process in parallel, and after that re-merge
every chunck taking in account the possibilities that a quote, a delimiter
or an end of line could be at an end of a chunk or at its beginning, but it
seems like not a good idea from what I read.

You could also take a look of what has been done in the ReadAsync method of
the csvHelper reading the next pages in this order:
1] discussion of the feature :
https://github.com/JoshClose/CsvHelper/issues/202
2] doc : https://joshclose.github.io/CsvHelper/reading/
3] test:
https://github.com/JoshClose/CsvHelper/blob/master/src/CsvHelper.Tests/Async/ReadingTests.cs

From my point of view, the multiline support should be done asynchronously
with the purpose of reading the file while processing it, but not chunking
it and readinh different parts in parallel ( some infos here :
https://stackoverflow.com/questions/41413626/is-there-a-way-to-use-parallel-processing-to-read-chunks-from-a-file-and-join-to
).

Hope that it helps, and really interesting by what you think about it ?!

*Bastien Besson* -* Freelance Technical Lead **.NET*

*Tel :* *06 20 07 87 48*
*Email :* *bastien.besson@gmail.com <bastien.besson@gmail.com>*

**Comment by bytefish** [comment]
I have created a develop branch, where I am going to experiment with some implementations. 

First of all I have simplified the Parser by translating Pythons CSV Parser:

* https://github.com/bytefish/TinyCsvParser/blob/develop/TinyCsvParser/TinyCsvParser/Parsers/CsvReader.cs

Going all async with it is simple, it is just a matter of using the StreamReader Async methods. But I think it is very far from trivial to parallelize the parsing, actually impossible. Chunking the file and grouping them again sounds too brittle to me. Without any judgement: If it is working for the CsvHelper library it is good for them, I don't want to maintain it.

So the problem with the above parser is, that it is inherently sequential by evaluating character by character. But does Tokenizing the Lines in parallel yield a huge benefit? Is the mapping more costly or the parsing?

I am also still thinking about: I have used PLINQ for the Parser, which doesn't work well with async / await. I used PLINQ, because it leads to quite simple pipelines for processing the CSV results in parallel and it was always a joy to work with. How would I build such Parallel Pipelines without using PLINQ? Then again do people actually use the PLINQ features or prefer async / await? 

I am still thinking about such things, that's why I am unsure in which direction to go.
